[BUG] Flaky `exec_command` output on SSH channel

##### SUMMARY
<!--- Explain the problem briefly below -->
`pylibsshext.channel.Channel.exec_command()` returns empty output from time to time. One invocation in tests returned an empty string for `exec_command('echo -n Hello World').stdout`.

##### ISSUE TYPE
- Bug Report

##### PYLISSH and LIBSSH VERSION
<!--- Paste verbatim output between quotes -->
N/A

##### OS / ENVIRONMENT
<!--- Provide all relevant information below, e.g. target OS versions, network device firmware, etc. -->
Currently observed in GHA under Python 3.5 + macOS. But it's not persistent so I assume that it may be reproducible in other envs.

##### STEPS TO REPRODUCE
<!--- Describe exactly how to reproduce the problem, using a minimal test-case -->

Restart `tests/unit/channel_test.py::test_exec_command` until it fails.

<!--- HINT: You can paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
<!--- Describe what you expected to happen when running the steps above -->
The test is always green.

##### ACTUAL RESULTS
<!--- Describe what actually happened. If possible run with extra verbosity (-vvvv) -->
The test is occasionally failing:

<!--- Paste verbatim command output between quotes -->
```python
=================================== FAILURES ===================================
______________________________ test_exec_command _______________________________
[gw1] darwin -- Python 3.5.9 /Users/runner/runners/2.263.0/work/pylibssh/pylibssh/.tox/test-binary-dists/bin/python

ssh_channel = <pylibsshext.channel.Channel object at 0x108861410>

    def test_exec_command(ssh_channel):
        """Test getting the output of a remotely executed command."""
        u_cmd_out = ssh_channel.exec_command('echo -n Hello World').stdout.decode()
>       assert u_cmd_out == u'Hello World'  # noqa: WPS302
E       AssertionError: assert '' == 'Hello World'
E         - Hello World

ssh_channel = <pylibsshext.channel.Channel object at 0x108861410>
u_cmd_out  = ''

tests/unit/channel_test.py:28: AssertionError
```

Ref: https://github.com/ansible/pylibssh/runs/724505447?check_suite_focus=true#step:10:212

--- a/src/pylibsshext/channel.pyx
+++ b/src/pylibsshext/channel.pyx
@@ -35,7 +35,7 @@ cdef int _process_outputs(libssh.ssh_session session,
                           void *data,
                           libssh.uint32_t len,
                           int is_stderr,
-                          void *userdata):
+                          void *userdata) with gil:
     if len == 0:
         return 0
     data_b = <bytes>(<char *>data)[:len]

--- a/src/pylibsshext/channel.pyx
+++ b/src/pylibsshext/channel.pyx
@@ -136,13 +136,13 @@ cdef class Channel:
         rc = libssh.ssh_channel_request_exec(self._libssh_channel, command.encode("utf-8"))
 
         if rc != libssh.SSH_OK:
-            libssh.ssh_channel_free(self._libssh_channel)
+            self.close()
             raise CalledProcessError()
+        result = CompletedProcess(args=command, returncode=-1, stdout=b'', stderr=b'')
 
         cdef callbacks.ssh_channel_callbacks_struct cb
         memset(&cb, 0, sizeof(cb))
         cb.channel_data_function = <callbacks.ssh_channel_data_callback>&_process_outputs
-        result = CompletedProcess(args=command, returncode=-1, stdout=b'', stderr=b'')
         cb.userdata = <void *>result
         callbacks.ssh_callbacks_init(&cb)
         callbacks.ssh_set_channel_callbacks(self._libssh_channel, &cb)
@@ -158,5 +158,6 @@ cdef class Channel:
 
     def close(self):
         if self._libssh_channel is not NULL:
+            libssh.ssh_channel_close(self._libssh_channel)
             libssh.ssh_channel_free(self._libssh_channel)
             self._libssh_channel = NULL

--- a/src/pylibsshext/channel.pyx
+++ b/src/pylibsshext/channel.pyx
@@ -61,8 +61,9 @@ cdef class Channel:
             raise LibsshChannelException("Failed to open_session: [%d]" % rc)
 
     def __dealloc__(self):
-        libssh.ssh_channel_free(self._libssh_channel)
-        self._libssh_channel = NULL
+        if self._libssh_channel is not NULL:
+            libssh.ssh_channel_free(self._libssh_channel)
+            self._libssh_channel = NULL
 
     def request_shell(self):
         self.request_pty()
@@ -133,17 +134,10 @@ cdef class Channel:
         return response
 
     def exec_command(self, command):
-        cdef libssh.ssh_channel channel = libssh.ssh_channel_new(self._libssh_session)
-        if channel is NULL:
-            raise MemoryError
-        rc = libssh.ssh_channel_open_session(channel)
-        if rc != libssh.SSH_OK:
-            libssh.ssh_channel_free(channel)
-            raise CalledProcessError()
+        rc = libssh.ssh_channel_request_exec(self._libssh_channel, command.encode("utf-8"))
 
-        rc = libssh.ssh_channel_request_exec(channel, command.encode("utf-8"))
         if rc != libssh.SSH_OK:
-            libssh.ssh_channel_free(channel)
+            libssh.ssh_channel_free(self._libssh_channel)
             raise CalledProcessError()
 
         cdef callbacks.ssh_channel_callbacks_struct cb
@@ -152,12 +146,11 @@ cdef class Channel:
         result = CompletedProcess(args=command, returncode=-1, stdout=b'', stderr=b'')
         cb.userdata = <void *>result
         callbacks.ssh_callbacks_init(&cb)
-        callbacks.ssh_set_channel_callbacks(channel, &cb)
+        callbacks.ssh_set_channel_callbacks(self._libssh_channel, &cb)
 
-        libssh.ssh_channel_send_eof(channel)
-        result.returncode = libssh.ssh_channel_get_exit_status(channel)
+        libssh.ssh_channel_send_eof(self._libssh_channel)
 
-        libssh.ssh_channel_free(channel)
+        result.returncode = self.get_channel_exit_status()
 
         return result
 
@@ -165,5 +158,6 @@ cdef class Channel:
         return libssh.ssh_channel_get_exit_status(self._libssh_channel)
 
     def close(self):
-        libssh.ssh_channel_free(self._libssh_channel)
-        self._libssh_channel = NULL
+        if self._libssh_channel is not NULL:
+            libssh.ssh_channel_free(self._libssh_channel)
+            self._libssh_channel = NULL

--- a/src/pylibsshext/channel.pyx
+++ b/src/pylibsshext/channel.pyx
@@ -62,8 +62,7 @@ cdef class Channel:
 
     def __dealloc__(self):
         if self._libssh_channel is not NULL:
-            libssh.ssh_channel_free(self._libssh_channel)
-            self._libssh_channel = NULL
+            self.close()
 
     def request_shell(self):
         self.request_pty()

--- a/src/pylibsshext/channel.pyx
+++ b/src/pylibsshext/channel.pyx
@@ -61,8 +61,7 @@ cdef class Channel:
             raise LibsshChannelException("Failed to open_session: [%d]" % rc)
 
     def __dealloc__(self):
-        if self._libssh_channel is not NULL:
-            self.close()
+        self.close()
 
     def request_shell(self):
         self.request_pty()

--- a/tests/unit/channel_test.py
+++ b/tests/unit/channel_test.py
@@ -29,7 +29,6 @@ def ssh_channel(ssh_client_session):
 @pytest.mark.xfail(
     reason='This test causes SEGFAULT, flakily. '
     'Ref: https://github.com/ansible/pylibssh/issues/57',  # noqa: WPS326
-    run=False,
     strict=False,
 )
 @pytest.mark.forked

--- a/tests/unit/channel_test.py
+++ b/tests/unit/channel_test.py
@@ -26,12 +26,6 @@ def ssh_channel(ssh_client_session):
         chan.close()
 
 
-@pytest.mark.xfail(
-    reason='This test causes SEGFAULT, flakily. '
-    'Ref: https://github.com/ansible/pylibssh/issues/57',  # noqa: WPS326
-    run=False,
-    strict=False,
-)
 @pytest.mark.forked
 def test_exec_command(ssh_channel):
     """Test getting the output of a remotely executed command."""

--- a/tests/unit/channel_test.py
+++ b/tests/unit/channel_test.py
@@ -26,6 +26,12 @@ def ssh_channel(ssh_client_session):
         chan.close()
 
 
+@pytest.mark.xfail(
+    reason='This test causes SEGFAULT, flakily. '
+    'Ref: https://github.com/ansible/pylibssh/issues/57',  # noqa: WPS326
+    run=False,
+    strict=False,
+)
 @pytest.mark.forked
 def test_exec_command(ssh_channel):
     """Test getting the output of a remotely executed command."""

